### PR TITLE
Bugfix/transient run fail 40

### PR DESCRIPTION
**Pull request by richpsharp**
Fixes unexpected behavior when the `transient_run=True` flag is used if a task had been previously run without that flag then the task would not reexecute even though the user would expect it to.

This PR adds a check for this state, modifies a test to cover this behavior and cleans up some of the wording in the docstring about this flag.

### CODE REVIEW

**Comment by richpsharp** [comment]
Hi @davemfish, I'm adding you as a reviewer because I think you are a TaskGraph fan and also this isn't a big PR at all.

File: tests/test_task.py [hunk 1]
```diff
@@ -1365,22 +1365,25 @@ def test_return_value_no_record(self):
...
         n_iterations = 3
+        task_graph = taskgraph.TaskGraph(self.workspace_dir, 0, 0)
         for iteration_id in range(n_iterations):
             transient_run = iteration_id == n_iterations-1
             LOGGER.debug(iteration_id)
-            task_graph = taskgraph.TaskGraph(self.workspace_dir, 0, 0)
             expected_value = 'a good value'
             value_task = task_graph.add_task(
                 func=_return_value_once,
                 transient_run=transient_run,
```
Comment (davemfish) [inline tests/test_task.py:1378]: This task does not re-execute on the last iteration when `transient_run=True`. The func would raise a `RuntimeError` if it did, right? Shouldn't `transient_run=True` always force an execution?

**Review by davemfish — COMMENTED** [review]
@richpsharp always a pleasure to review `Taskgraph`! I have a question about some docstring wording and also a question about one of the tests. They might end being the same question...where I'm not really understanding circumstances when a `transient_run=True` should _not_ re-execute.

File: taskgraph/Task.py [hunk 2]
```diff
@@ -628,11 +628,11 @@ def add_task(
             hardlink_allowed (bool): if ``copy_duplicate_artifact`` is True,
                 this will allow a hardlink rather than a copy when needed.
             transient_run (bool): if True a call with an identical execution
-                hash will be reexecuted on a subsequent instantiation of a
-                future TaskGraph object. If a duplicate task is submitted
+                hash will be reexecuted on the same or subsequent instantiation
+                of a TaskGraph object. If a duplicate task is submitted
                 to the same object it will not be re-run in any scenario.
```
[the remark this answers is not in the document]
Comment (richpsharp) [reply taskgraph/Task.py:633]: Great point! And thanks for pointing this out because it helps me get it straight in my head. Abstractly, Taskgraph is a dependency graph and an instance of a TaskGraph object traverses that dependency graph executing Tasks as allowed and needed. My intent for `transient_run` is that every time a Task is iterated over in the dependency graph (the equivalent of an instantiation of TaskGraph), that Task be executed. 

The confusing stuff about duplicate Tasks may be me over engineering TaskGraph. Since TaskGraph's abstract model is a graph, it's the equivalent of a set, it doesn't make sense to have duplicate nodes because it's the same node. In practice you'd have a similar confusion -- why indicate that a Task exists twice? So TaskGraph does detect this issue and rather than raise an error it logs a warning and discards the Task. I added this because I was anticipating, and use, cases where I'm not sure if a Task has been added to the graph yet and TaskGraph otherwise has no way of querying its state.

SO that's all coming down to this mishmash of docstring that I wrote. I will rephrase it to follow the abstraction here.

File: tests/test_task.py (hunk 1 above)
Comment (richpsharp) [reply tests/test_task.py:1378]: Now this one IS intentional. Given my comment above, it doesn't make sense to add duplicate Tasks to TaskGraph and the way it's set up is that if you do it doesn't break anything and instead logs an error. So this bit here is just ensuring that multiple transient run tasks in the same TaskGraph instance are only executed once. I thought that behavior was worth explicitly testing since I do it with non transient tasks elsewhere in the suite.

File: taskgraph/Task.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (richpsharp) [reply taskgraph/Task.py]: Okay I updated that docstring in 34416c438141c99d9ad93db7f58fa4cd1f36858c

**Comment by richpsharp** [comment]
Hi @davemfish, thanks for pushing on that docstring description. I think I made it better.

**Review by davemfish — APPROVED** [review]
@richpsharp this new docstring is very readable and helpful! It's clear what I should expect from a `transient_run` and why I might want that. Thanks also for describing the data structures. It's clear to me now exactly what a "duplicate" task implies. I'm not sure I've encountered it yet in practice, but the way it's handled (log but not crash) sounds very reasonable.
